Feat/new toast

ethanshar opened the pull request:
## Description
**Please Review - Don't Merge!**
_This contains lots of files, most of them are assets and generated types. It'll be best to review this PR by commits_ 

This is the new implementation of our Toast component (taken from the private lib)
The component's code has been refactored to a functional component 
Im using separate helper hooks for different parts of logic of the component. 

Few alterations I did
- I added support for `offline` preset for the private usage we have (in private we deepened on the `isConnected` prop)
- I added support for containerStyle, messageStyle, iconColor to give more customize capabilities to our usets

## Changelog
New Toast component (under Incubator) with better UI and lots of new features

Inbal-Tish commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,276 @@
+    testID
+  } = props;
+
+  const directions = useRef([
```
Why do you need useRef on Array if you're not modifying it anywhere?

ethanshar replied:
Cause we're passing the directions to PanView and we don't want to create a new ref for the array on each render.

Inbal-Tish commented on src/incubator/toast/helpers/useToastPresets.ts:
```diff
@@ -0,0 +1,75 @@
+const infoIcon = require('../assets/info.png');
+const redCloudIcon = require('../assets/redCloud.png');
+
+export default ({preset, icon, iconColor, message}: Pick<ToastProps, 'preset' | 'icon' | 'message' | 'iconColor'>) => {
```
Why do you use `Pick<ToastProps, ...>` when you extract the props?

ethanshar replied:
Because I want the hook to have an accurate typecheck. 
Instead of passing all props, I declare only the ones relevant to this hook

Inbal-Tish commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,276 @@
+  };
+
+  const renderToast = () => {
+    if (!visible && !isAnimating) {
```
Why this check is here as well? Is there a scenario where it will not return on line 209 but will return here?

ethanshar replied:
You are right, It was copied like that from the private code and I see that we have the same issue there as well.

Inbal-Tish commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,276 @@
+
+  const onDismiss = useCallback(() => {
+    clearTimer();
+    props.onDismiss?.();
```
Better use 'onDismiss' directly and not as 'props.onDismiss'

ethanshar replied:
fixed

lidord-wix reviewed (commented):
Looks good, the hooks are making a lot of sense :) I left few specific comments. 
The following are more generic:
1. It might be a bug in the example screen, but when dismissing the toast by sliding it down or up, you should press the `Toggle toast` button 3 times to present it again (instead of 1).
2. There's a bug in the attachment feature, you can see it when sliding the toast right/left (happens on private as well).
3. I think it is a better API to have a `customContent` prop instead of getting it as children, it'll also be more clear to us in the future.
4. Consider exporting the toast presets so users will be able to use `Toast.presets.SUCCESS`.

M-i-k-e-l commented on src/incubator/toast/helpers/useToastAnimation.ts:
```diff
@@ -0,0 +1,64 @@
+    }).start(_onAnimationEnd);
+
+    if (preset === ToastPreset.FAILURE && show) {
+      HapticService.triggerHaptic(HapticType.impactMedium, 'Toast');
```
I think this is a problem, the Haptic is an optional dependency, and the toast should be more generic.
We should either test here or create a function (or add an option to the current `triggerHaptic`) to only trigger if it's available and not trigger an error.

ethanshar replied:
Yes, you are right. 
I added a dedicate prop for that (`enableHapticFeedback`) and use that instead. 
In private we should enable it based on our logic

M-i-k-e-l commented on src/incubator/toast/helpers/useToastAnimation.ts:
```diff
@@ -0,0 +1,64 @@
+    outputRange: [positionMultiplier * toastHeight, 0]
+  });
+
+  return {isAnimating, toggleToast, toastOpacity, toastTranslateY};
```
Wouldn't exporting `opacityStyle` and `translationStyle` will be a little nicer? I mean to export them like this:
`{opacity: toastOpacity}` and `{transform: [{translateY: toastTranslateY}]}`.
And maybe `toggleToast` can be simply `toggle`.

ethanshar replied:
I agree. 
I changed to opacityStyle and translateStyle. 
As for the `toggleToast`, I actually like that it's explicit, I find it more readable

M-i-k-e-l commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,273 @@
+        <View
+          animated
+          useSafeArea
+          style={[{opacity: toastOpacity}, containerStyle]}
```
Shouldn't this be memoized?

ethanshar replied:
Fixed

M-i-k-e-l commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,273 @@
+  const renderRightElement = () => {
+    // NOTE: order does matter
+    if (showLoader) {
+      const loaderColors = [Colors.rgba(Colors.white, 0.3), Colors.rgba(Colors.grey30, 0.7)];
```
Is it better to keep this here or move it to a global const?

ethanshar replied:
I actually removed it all together, it belong to the private loader anyway

ethanshar commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,273 @@
+  const viewRef = useRef();
+  const [toastHeight, setToastHeight] = useState(500);
+
+  const {clearTimer, setDismissTimer} = useToastTimer(props);
+  const {icon, iconColor, accessibilityMessage} = useToastPresets(props);
```
Readability wise, I think it's better.. 
If I'll extract the relevant props at the top I'll have to do something like this
```
  const {icon: presetIcon, iconColor: presetIconColor, accessibilityMessage} = useToastPresets({preset, icon, iconColor});
```

I'll need to give everything new names, not sure if it's more readable, WDYT?

lidord-wix replied:
I think it even makes more sense that way, cause you understand that the iconColor, for example, has been changed by this hook.. and you also know exactly which props are affecting this hook. 
But it's up to you :)

ethanshar replied:
Ok, I changed it, tell me what you think..

lidord-wix replied:
Looks great :)

ethanshar commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,273 @@
+  };
+
+  const {isAnimating, toggleToast, toastOpacity, toastTranslateY} = useToastAnimation({
+    ...props,
```
Here's it's easier cause Im returning different stuff.. 
Fixed

ethanshar commented on src/incubator/toast/index.tsx:
```diff
@@ -0,0 +1,273 @@
+
+  const renderIcon = () => {
+    const tintColor = iconColor;
+    return <Image source={icon} resizeMode={'contain'} style={styles.icon} tintColor={tintColor}/>;
```
Definitely! Fixed 👌

ethanshar commented on src/incubator/toast/helpers/useToastPresets.ts:
```diff
@@ -0,0 +1,75 @@
+import {Colors} from 'style';
+import {ToastProps, ToastPreset} from '../types';
+
+const checkMarkIcon = require('../assets/checkmarkFlat.png');
+const exclamationIcon = require('../assets/exclamationFill.png');
+const infoIcon = require('../assets/info.png');
+const redCloudIcon = require('../assets/redCloud.png');
```
Im considering these assets part of the component so I added them under the component asset folder.

ethanshar commented on src/incubator/toast/helpers/useToastPresets.ts:
```diff
@@ -0,0 +1,75 @@
+const infoIcon = require('../assets/info.png');
+const redCloudIcon = require('../assets/redCloud.png');
+
+export default ({preset, icon, iconColor, message}: Pick<ToastProps, 'preset' | 'icon' | 'message' | 'iconColor'>) => {
```
What do you mean? can you elaborate..

M-i-k-e-l replied:
Something like this (see `useBannerPresets` in our private lib):
```
const toastPresets = {};
toastPresets[ToastPreset.SUCCESS] = {
  icon: checkMarkIcon,
  iconColor: icon || iconColor ? iconColor : Colors.green40,
  presetText: 'Success'
};
...
return toastPresets;
```

ethanshar replied:
Ok, I did something similar

ethanshar commented on src/incubator/toast/helpers/useToastTimer.ts:
```diff
@@ -0,0 +1,21 @@
+export default ({autoDismiss, onDismiss}: Pick<ToastProps, 'autoDismiss' | 'onDismiss'>) => {
+  const timer = useRef<ReturnType<typeof setTimeout>>();
+
+  const clearTimer = useCallback(() => {
```
Yes, you are right. It's a leftover from the previous code. 
Fixed 👍

ethanshar commented on src/incubator/toast/helpers/useToastAnimation.ts:
```diff
@@ -0,0 +1,64 @@
+interface UseToastAnimationProps extends Pick<ToastProps, 'visible' | 'preset' | 'position' | 'onAnimationEnd'> {
+  toastHeight: number;
+  playAccessibilityFeatures: () => void;
+  setDismissTimer: () => void;
```
We only use this type here and it's pretty simple type.
If it was used in other places then maybe.. WYDT? I don't mind either way

M-i-k-e-l replied:
`setTimer` is used in both `useToastAnimation` and `useToastTimer`, but it's not a big deal.

ethanshar replied:
It's exported in `useToastTimer`, but we declare its typing once in `useToastAnimation`

lidord-wix commented:
> Looks good, the hooks are making a lot of sense :) I left few specific comments. The following are more generic:
> 
> 1. It might be a bug in the example screen, but when dismissing the toast by sliding it down or up, you should press the `Toggle toast` button 3 times to present it again (instead of 1).
> 2. There's a bug in the attachment feature, you can see it when sliding the toast right/left (happens on private as well).
> 3. I think it is a better API to have a `customContent` prop instead of getting it as children, it'll also be more clear to us in the future.
> 4. Consider exporting the toast presets so users will be able to use `Toast.presets.SUCCESS`.

@ethanshar did you fix these issues?

M-i-k-e-l commented on src/incubator/toast/index.tsx:
```diff
@@ -165,8 +168,8 @@ const Toast = (props: PropsWithChildren<ToastProps>) => {
   const renderIcon = () => {
-    const tintColor = iconColor;
-    return <Icon source={icon} resizeMode={'contain'} style={styles.icon} tintColor={tintColor}/>;
+    const tintColor = toastPreset.iconColor;
+    return <Icon source={toastPreset.icon} resizeMode={'contain'} style={styles.icon} tintColor={tintColor}/>;
```
Are you keeping the `const tintColor` for readability?

ethanshar replied:
Nope (: 
Forgot it when I refactored the code. Fixed!

ethanshar commented:
> > > Looks good, the hooks are making a lot of sense :) I left few specific comments. The following are more generic:
> > > 
> > > 1. It might be a bug in the example screen, but when dismissing the toast by sliding it down or up, you should press the `Toggle toast` button 3 times to present it again (instead of 1).
> > > 2. There's a bug in the attachment feature, you can see it when sliding the toast right/left (happens on private as well).
> > > 3. I think it is a better API to have a `customContent` prop instead of getting it as children, it'll also be more clear to us in the future.
> > > 4. Consider exporting the toast presets so users will be able to use `Toast.presets.SUCCESS`.
> > 
> > 
> > @ethanshar did you fix these issues?
> 
> Not all, I'll go over them again.. thanks

@lidord-wix 
1. The first issue is more of an example screen issue, anyway I fixed it, you can take a look. 
2. Fixed
3. I prefer not to change API, I copied the code the private and I'd like to have a smooth transition to the pubic implementation without having a big migration
4. Done!
